### ISSUE TITLE
[42] Panoptic SegFormer: Delving Deeper into Panoptic Segmentation with Transformers

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#70]
[paper](https://arxiv.org/pdf/2109.03814.pdf)
[code](https://github.com/zhiqi-li/Panoptic-SegFormer)

- Pyramid Vision Transformer(PVT) v2 백본 (#69)
- multiscale feature를 deformable attention encoder로 결합
- location decoder로 각 instance의 위치 정보를 찾아낸 다음 이 정보를 mask decoder에 쿼리로 입력해 마스크를 뽑는 방식

# Introduction
![image](https://user-images.githubusercontent.com/16400591/146123497-f474b2f9-19c5-4a61-a81c-7ca158e7f922.png)

# Methods
## 3.1. Overall Architecture
![image](https://user-images.githubusercontent.com/16400591/146123463-7c9d29b3-fb42-4277-80cd-402c224fc329.png)
1. backbone feature 들이 flatten 되어서 encoder 거침 (C3, C4, C5 는 각각 1/8, 1/16, 1/32 크기)
2. ( thing query, (1) output ) 를 input 으로 받아서, location decoder 에서 bbox 찾아냄
3. ( stuff query, (2) output, (1) output ) 를 input으로 받아서 mask decdoer 돌림

C3, C4, C5 를 flatten 시켜서 FC 를 하나 태우는데, channel size 는 256 으로 고정한다.
결과적으로 `L1 x 256`, `L2 x 256`, `L3 x 256` 크기의 맵들이 나온다.
`L_i` 는 다음과 같이 정의된다.
![image](https://user-images.githubusercontent.com/16400591/146135460-80568ae6-8112-4d0a-adc3-2ddf3da59175.png)

## 3.2. Transformer Encoder
Deformable DETR 에서 제안한 encoding 방법 사용.

### 3.3.1 Query Decoupling Strategy
지금까지처럼 `things query`와 `stuffs query`를 하나에 담는 건 비효율적이라 주장함.
좀 찾아봤는데, COCO 기준 아래 라벨맵에서 앞 80개가 things, 뒤 53개가 stuff 다.
https://github.com/zhiqi-li/Panoptic-SegFormer/blob/38b0b46f6e36e0dac7c49065a28fdbd03ff29e9b/easymd/models/utils/visual.py#L90-L113

thing query 는 300개
stuff query 는 데이터셋마다 다름.

![image](https://user-images.githubusercontent.com/16400591/146129937-fc563b7f-abe5-444c-a92a-9b9b4e2564ce.png)
보라색 thing-query 에 대해서는 bipartite matching 을 주고,
location 된 stuff-query 들은 fixed-matching 을 수행함.

### 3.3.2 Location Decoder
location decoder는 Deformable DETR과 동일하다.
training 할 때는 location decoder 에 auxiliary MLP head 붙여서 Detection Loss `L_{det}` 추가로 뽑아냄.

location decoder는 꼭 box 를 만들 필요는 없음.
object의 center of mass 를 예측하게도 만들어 봤는데, box 모델이랑 비슷한 성능을 내더라.

### 3.3.3 Mask Decoder
![image](https://user-images.githubusercontent.com/16400591/146139794-488ef982-b884-4eb6-a95c-2333a7016def.png)

overall architecture 이미지의 (d) 에 해당하는 부분.
segmentation mask 를 예측해야 해서 transformer decoder 날 것 그대로 사용했음.

thing query 를 mask decoder 에 통과시켜서 모든 "thing category" 를 예측해야 함.
그니까 그냥 segmentation 하듯이 prob map 을 예측하면 되는 것.
category 별 pixel-wise logit 뽑고 Cross-Entropy 쓰면 됨.

sutff query 는 stuff 들만 예측하면 된다.
그니까 애초에 stuff 개수가 정해져 있으니 (coco 기준 53개) 그만큼의 query를 만들고 각각에 대한 probability map을 뽑아내는 방식.

DETR 논문에서 attention 때문에 수렴이 오래걸린다 해서 2가지 테크닉 적용.
1. ultra-light FC head 사용 (200 parameter)
2. Attention map supervision (attention map 이 mask 에 의해 supervised 됨. learning process 가속화)

mask 를 이용해서 attention map에 deep supervision을 준다? 무슨 소리고... 하니,
![image](https://user-images.githubusercontent.com/16400591/146139879-18bb74fe-4bd0-4ab4-9e9a-906de2b5f625.png)
attention map A의 총 크기는 `N×h×(L1+L2+L3)`.
N 은 `N_{thing} + N_{stuff}` 이다. coco 기준 353.
h는 attention head 수.
L1, L2, L3 는 encoder 쪽(위쪽) 에 설명해 놨음.
이 attention 들을 (`N×h×L1`, `N×h×L2`, `N×h×L3`) bilinear resize 로 크기 맞춰주고,
FC 하나 통과시켜서 mask 예측하게 학습함.

## 3.4. Loss Function
![image](https://user-images.githubusercontent.com/16400591/146131193-a64f12e9-f726-40b3-adce-6fbba152b57f.png)
`λ_{things}`, `λ_{stuff}` 는 이미지 별로 portion을 보고 조정되었음. 합은 1.

![image](https://user-images.githubusercontent.com/16400591/146137102-de306152-c73f-412e-ad73-14587185918f.png)
![image](https://user-images.githubusercontent.com/16400591/146137316-c1170f23-6d82-4639-9b3f-778bffa2d55b.png)
`λ_{cls}`, `λ_{seg}`, `λ_{det}` 는 각각 2, 1, 1
`L_{det}` 는 Deformable DETR loss, location decoder에 붙음
`L_{seg}` 는 dice loss
`L_{cls}` 는 focal loss

## 3.5. Mask-Wise Merging Inference
pixel 별로 정답을 맞추는게 panoptic segmentation 인데,
pixel 별 argmax 로 하나 할당시키니까 false positive 가 너무 많더라.

제안하는 방법은 다음과 같다.
먼저 confidence score s 를 구한다.
![image](https://user-images.githubusercontent.com/16400591/146137678-b08a2c9a-f76a-4ad1-9303-a9e85f6178c4.png)
`p_i` 는 i번째 result 의 class probability.
`m_i[h, w]` 는 [h, w] 위치의 mask logit.
α, β 는 classification probability, segmentation qualities 를 weighting 하는 parameter.
수식에서 볼 수 있듯이 mask prob 이 0.5 이상인 것만 사용함

이렇게 구한 confidence score 가 낮은 애들은 그냥 pred를 안하는 방식.
mask prob 도 너무 낮으면 아예 pred 를 안함.
![image](https://user-images.githubusercontent.com/16400591/146130141-0f56a57d-266a-4fc0-b6ff-b31e47ab4031.png)

# Result
십자가 표시(†)는 pretrain imagenet-22K

### COCO val 2017
![image](https://user-images.githubusercontent.com/16400591/146120742-71b993a8-e5b4-4c17-9982-867d46e25e88.png)
![image](https://user-images.githubusercontent.com/16400591/146120896-bd2084ee-fec0-4d9d-9372-e398c33e4562.png)

### COCO test-dev
![image](https://user-images.githubusercontent.com/16400591/146120986-116c3164-1067-4d1c-b09a-98f0e261173a.png)

### ADE20K val
![image](https://user-images.githubusercontent.com/16400591/146121099-16b29b65-0245-4d2f-ad3e-b42d0938cc6b.png)

## Ablations
Panoptic SegFormer (R50) 모델 사용. 
### Effect of Deformation Attention
deformable attention 쓴 게 더 좋았음.
특히 multi-scale feature를 쓰는 게 성능향상에 큰 도움.
![image](https://user-images.githubusercontent.com/16400591/146138924-e367bcd4-f327-48fe-8ab2-059ea9140ad5.png)

### Mask-wise Merging
(p) 가 pixel-wise argmax 방법
(p*)은  [MaskFormer](https://arxiv.org/pdf/2107.06278.pdf) 에서 제안한 방법 (이거도 mask prob 까지 신경써줌)
제안한 (m) 이 제일 좋음 
성능은 1% 정도 오름
![image](https://user-images.githubusercontent.com/16400591/146139004-cbe6b9ef-a587-4659-a098-95321f75ce8c.png)

### Effect of Location Decoder
box-free 모델 흠터레스팅!
![image](https://user-images.githubusercontent.com/16400591/146139416-92d0895d-0289-4508-acf2-5b46794ef512.png)

### Mask Decoder
decoder 를 3개만 써도 나쁘지 않음
![image](https://user-images.githubusercontent.com/16400591/146140046-ab1a550f-5c66-4ca9-88d7-816d9a321621.png)

### Effect of Query Decoupling Strategy
대략 성능상 1% 정도 이득
![image](https://user-images.githubusercontent.com/16400591/146139487-c31ab157-9085-46f9-9ab8-65d76637163b.png)

# Conclusion
Limitation
1. deformable attention을 써서 속도가 느림
2. 크거나 작은 녀석들에 대해서는 잘 동작하지 않음
